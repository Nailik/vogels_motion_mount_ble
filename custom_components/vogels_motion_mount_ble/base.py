"""Base entity which all other entity platform classes can inherit.

As all entity types have a common set of properties, you can
create a base entity like this and inherit it in all your entity platforms.

This just makes your code more efficient and is totally optional.

See each entity platform (ie sensor.py, switch.py) for how this is inheritted
and what additional properties and methods you need to add for each entity type.

"""

import logging
from typing import Any

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import ExampleCoordinator

_LOGGER = logging.getLogger(__name__)


class ExampleBaseEntity(CoordinatorEntity):
    """Base Entity Class.

    This inherits a CoordinatorEntity class to register your entites to be updated
    by your DataUpdateCoordinator when async_update_data is called, either on the scheduled
    interval or by forcing an update.
    """

    coordinator: ExampleCoordinator

    # ----------------------------------------------------------------------------
    # Using attr_has_entity_name = True causes HA to name your entities with the
    # device name and entity name.  Ie if your name property of your entity is
    # Voltage and this entity belongs to a device, Lounge Socket, this will name
    # your entity to be sensor.lounge_socket_voltage
    #
    # It is highly recommended (by me) to use this to give a good name structure
    # to your entities.  However, totally optional.
    # ----------------------------------------------------------------------------
    _attr_has_entity_name = True

    def __init__(self, coordinator: ExampleCoordinator) -> None:
        """Initialise entity."""
        super().__init__(coordinator)

    @property
    def device_info(self) -> DeviceInfo:
        """Return device information."""

        # ----------------------------------------------------------------------------
        # Identifiers are what group entities into the same device.
        # If your device is created elsewhere, you can just specify the indentifiers
        # parameter to link an entity to a device.
        # If your device connects via another device, add via_device parameter with
        # the indentifiers of that device.
        #
        # Device identifiers should be unique, so use your integration name (DOMAIN)
        # and a device uuid, mac address or some other unique attribute.
        # ----------------------------------------------------------------------------
        return DeviceInfo(
            name=self.coordinator.name,
            manufacturer="Vogelsr",
            model="Motion Mount",
            identifiers={(DOMAIN, self.coordinator.mac)},
        )

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return self.coordinator.name

    @property
    def unique_id(self) -> str:
        """Return unique id."""

        # ----------------------------------------------------------------------------
        # All entities must have a unique id across your whole Home Assistant server -
        # and that also goes for anyone using your integration who may have many other
        # integrations loaded.
        #
        # Think carefully what you want this to be as changing it later will cause HA
        # to create new entities.
        #
        # It is recommended to have your integration name (DOMAIN), some unique id
        # from your device such as a UUID, MAC address etc (not IP address) and then
        # something unique to your entity (like name - as this would be unique on a
        # device)
        #
        # If in your situation you have some hub that connects to devices which then
        # you want to create multiple sensors for each device, you would do something
        # like.
        #
        # f"{DOMAIN}-{HUB_MAC_ADDRESS}-{DEVICE_UID}-{ENTITY_NAME}""
        #
        # This is even more important if your integration supports multiple instances.
        # ----------------------------------------------------------------------------
        return f"{DOMAIN}-{self.coordinator.mac}"
