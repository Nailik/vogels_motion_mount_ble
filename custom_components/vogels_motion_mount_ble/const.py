"""Constants for the Vogels Motion Mount BLE integration."""

DOMAIN = "vogels_motion_mount_ble"

SERVICE_UUID = "3e6fe65d-ed78-11e4-895e-00026fd5c52c"

CHAR_NAME_UUID = "c005fa37-0651-4800-b000-000000000000"
CHAR_SIZE_UUID = "c005fa2b-0651-4800-b000-000000000000"
CHAR_DISTANCE_UUID = "c005fa00-0651-4800-b000-000000000000"
CHAR_ROTATION_UUID = "c005fa01-0651-4800-b000-000000000000"
CHAR_AUTOMATIC_UUID = "c005fa02-0651-4800-b000-000000000000"
CHAR_FREEZE_UUID = "c005fa14-0651-4800-b000-000000000000"
CHAR_PRESET_UUIDS = [
    "c005fa0a-0651-4800-b000-000000000000",
    "c005fa0b-0651-4800-b000-000000000000",
    "c005fa0c-0651-4800-b000-000000000000",
    "c005fa0d-0651-4800-b000-000000000000",
    "c005fa0e-0651-4800-b000-000000000000",
    "c005fa0f-0651-4800-b000-000000000000",
    "c005fa10-0651-4800-b000-000000000000",
]
CHAR_PRESET_UUID = "c005fa2a-0651-4800-b000-000000000000"
