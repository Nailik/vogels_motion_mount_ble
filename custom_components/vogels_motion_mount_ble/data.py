"""Holds the data that is stored in memory by the Vogels Motion Mount Integration."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class VogelsMotionMountPinSettings(Enum):
    """Pin settings."""

    Deactivated = 12
    Single = 13
    Multi = 15


class VogelsMotionMountAuthenticationType(Enum):
    """Current authentication status."""

    Wrong = 0
    Control = 1
    Full = 2


class VogelsMotionMountAutoMoveType(Enum):
    """Authentication options."""

    Hdmi_1_On = 0
    Hdmi_1_Off = 1
    Hdmi_2_On = 4
    Hdmi_2_Off = 5
    Hdmi_3_On = 8
    Hdmi_3_Off = 9
    Hdmi_4_On = 12
    Hdmi_4_Off = 13
    Hdmi_5_On = 16
    Hdmi_5_Off = 17

    def is_on(self) -> bool:
        """Return True if this enum represents an 'On' state."""
        return self.value % 4 == 0


@dataclass
class VogelsMotionMountAuthenticationStatus:
    """Current authentication status."""

    auth_type: VogelsMotionMountAuthenticationType
    cooldown: int | None = None


@dataclass
class VogelsMotionMountPreset:
    """Preset data."""

    index: int
    data: VogelsMotionMountPresetData | None


@dataclass
class VogelsMotionMountPresetData:
    """Preset data."""

    distance: int
    name: str
    rotation: int


@dataclass
class VogelsMotionMountMultiPinFeatures:
    """Current set of features for authorised user."""

    change_default_position: bool
    change_name: bool
    change_presets: bool
    change_tv_on_off_detection: bool
    disable_channel: bool
    start_calibration: bool


@dataclass
class VogelsMotionMountVersions:
    """Version data."""

    ceb_bl_version: str
    mcp_bl_version: str
    mcp_fw_version: str
    mcp_hw_version: str


@dataclass
class VogelsMotionMountData:
    """Holds the data of the device."""

    automove: VogelsMotionMountAutoMoveType
    connected: bool
    distance: int
    freeze_preset_index: int
    multi_pin_features: VogelsMotionMountMultiPinFeatures
    name: str
    pin_setting: VogelsMotionMountPinSettings
    presets: list[VogelsMotionMountPreset]
    rotation: int
    tv_width: int
    versions: VogelsMotionMountVersions
    permissions: VogelsMotionMountPermissions
    requested_distance: int | None = None
    requested_rotation: int | None = None


@dataclass
class VogelsMotionMountPermissions:
    """Permissions for currently used pin."""

    auth_status: VogelsMotionMountAuthenticationStatus
    change_settings: bool
    change_default_position: bool
    change_name: bool
    change_presets: bool
    change_tv_on_off_detection: bool
    disable_channel: bool
    start_calibration: bool
