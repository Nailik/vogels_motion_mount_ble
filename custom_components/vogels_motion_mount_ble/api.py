"""API Placeholder.

You should create your api seperately and have it hosted on PYPI.  This is included here for the sole purpose
of making this example code executable.
"""

import asyncio
from dataclasses import dataclass, replace
import logging
from .const import CHAR_DISTANCE_UUID, CHAR_ROTATION_UUID, CHAR_PRESET_UUID
from bleak import BleakClient, BleakScanner

_LOGGER = logging.getLogger(__name__)

@dataclass
class VogelsMotionMountData:
    """Class to hold api data."""

    connected: bool = False
    distance: int | None = None
    rotation: int | None = None

class API:
    """Class for example API."""

    def __init__(self, mac: str, pin: str | None, coordinator) -> None:
        """Initialise."""
        self._mac = mac
        self._pin = pin
        self._coordinator = coordinator
        self._data = VogelsMotionMountData(connected=False)
        self._client: BleakClient | None = None
        self._disconnected_event = asyncio.Event()

    async def testConnect(self):
        """Test connection to the BLE device once using BleakClient."""
        try:
            _LOGGER.debug("start test connection")
            async with BleakClient(self._mac, timeout=120) as client:
                _LOGGER.debug("Connected to device: %s", client.is_connected)
        except Exception as err:
            _LOGGER.error("Failed to connect to device: %s", err)
            raise APIConnectionError("Error connecting to api.") from err


    async def perform_action(self):
        _LOGGER.exception("write_gatt_char")
        result = await self._client.write_gatt_char(CHAR_PRESET_UUID, bytes([1]), response=True)
        _LOGGER.exception("write_gatt_char result: %s", result)

    def update(self, **kwargs):
        """Update one or more fields, retaining others from existing data. Then notify the coordinator."""
        self._data = replace(self._data, **kwargs)
        self._coordinator.async_set_updated_data(self._data)

    def _handle_disconnect(self, _):
        _LOGGER.debug("handle Device disconnected!")
        self._disconnected_event.set()

    def _handle_distance_change(self, _, data):
        _LOGGER.debug("distance change %s", data)
        self.update(distance=int.from_bytes(data, "little"))

    def _handle_rotation_change(self, _, data):
        _LOGGER.debug("rotation change %s", data)
        self.update(rotation=int.from_bytes(data, "little"))

    # todo make it optional if the connection should be maintained or connect on command (when required to send command) or poll time
    async def maintainConnection(self):
        """Maintain connection to device."""
        while True:
            _LOGGER.debug("scanning for device %s", self._mac)
            device = await BleakScanner.find_device_by_address(self._mac)

            if device is None:
                _LOGGER.debug("no device %s found, wait then scan again", self._mac)
                await asyncio.sleep(5)
                continue

            try:
                _LOGGER.debug("connecting to device")
                async with BleakClient(device, disconnected_callback=self._handle_disconnect, timeout=120) as client:
                    self.connected = client.is_connected
                    self._client = client
                    _LOGGER.debug("maintain connected to device %s", self._client.is_connected)
                    self.update(connected = self._client.is_connected)

                    self._handle_distance_change(None, await client.read_gatt_char(CHAR_DISTANCE_UUID))
                    self._handle_rotation_change(None, await client.read_gatt_char(CHAR_ROTATION_UUID))

                    await client.start_notify(CHAR_DISTANCE_UUID, self._handle_distance_change)
                    await client.start_notify(CHAR_ROTATION_UUID, self._handle_rotation_change)

                    await self._disconnected_event.wait()
                    # reset event
                    self._disconnected_event.clear()
                    _LOGGER.debug("device disconnected %s", self.connected)
                    self.update(connected = client.is_connected)
                    await asyncio.sleep(1)
            except Exception:
                # catch bleak.exc.BleakError: No backend with an available connection slot that can reach address D9:13:5D:AB:3B:37 was found
                _LOGGER.exception("Exception while connecting/connected")
                await asyncio.sleep(5)


class APIConnectionError(Exception):
    """Exception class for connection error."""
