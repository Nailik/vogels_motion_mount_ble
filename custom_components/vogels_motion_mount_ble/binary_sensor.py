"""Interfaces with the Integration 101 Template api sensors."""

import logging

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from . import MyConfigEntry
from .base import ExampleBaseEntity
from .const import DOMAIN
from homeassistant.helpers import config_validation as cv
import voluptuous as vol
from homeassistant.helpers.typing import ConfigType
from homeassistant.core import HomeAssistant, ServiceCall, callback

_LOGGER = logging.getLogger(__name__)

async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: MyConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the Binary Sensors."""
    coordinator = hass.data[DOMAIN][config_entry.entry_id]
    async_add_entities([ExampleBinarySensor(coordinator)])


class ExampleBinarySensor(ExampleBaseEntity, BinarySensorEntity):
    """Implementation of a sensor."""

    _attr_device_class = BinarySensorDeviceClass.CONNECTIVITY
    _attr_has_entity_name = True
    _attr_translation_key = "connection"
    _attr_should_poll = True

    @callback
    def _handle_coordinator_update(self) -> None:
        """Update sensor with latest data from coordinator."""
        _LOGGER.exception("_handle_coordinator_update binary sensor")
        # This method is called by your DataUpdateCoordinator when a successful update runs.
        self.async_write_ha_state()

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "connection"

    @property
    def unique_id(self) -> str:
        """Return unique id."""
        # All entities must have a unique id.  Think carefully what you want this to be as
        # changing it later will cause HA to create new entities.
        return f"{DOMAIN}-{self.coordinator.mac}-connected"

    @property
    def is_on(self) -> bool:
        """Return if the binary sensor is on."""
        _LOGGER.exception("is_on binary sensor")
        if not self.coordinator.data:
            _LOGGER.error("No data available in coordinator")
            return None
        # This needs to enumerate to true or false
        return self.coordinator.data.connected

