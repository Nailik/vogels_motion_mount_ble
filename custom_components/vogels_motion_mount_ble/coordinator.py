"""Integration 101 Template integration using DataUpdateCoordinator."""

import logging

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PIN
from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

from .api import API

_LOGGER = logging.getLogger(__name__)

class ExampleCoordinator(DataUpdateCoordinator):
    """My example coordinator."""

    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
        """Initialize coordinator."""

        # Set variables from values entered in config flow setup
        self.mac = config_entry.data[CONF_HOST]
        self._name = config_entry.data[CONF_NAME]
        self._pin = config_entry.data.get(CONF_PIN)

        # Initialise DataUpdateCoordinator (that's the device name shown to the user)
        super().__init__(
            hass,
            _LOGGER,
            name=config_entry.title,
            config_entry=config_entry,
        )

        # Initialise your api here
        self.api = API(mac=self.mac, pin=self._pin, coordinator=self)
        hass.loop.create_task(self.api.maintainConnection())
