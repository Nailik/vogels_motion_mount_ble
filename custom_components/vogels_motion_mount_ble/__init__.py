"""Integration for a Vogels Motion Mount via BLE."""

from __future__ import annotations

import logging
from homeassistant.const import __version__ as ha_version
from packaging import version
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import (
    ConfigEntryAuthFailed,
    ConfigEntryError,
    ConfigEntryNotReady,
)

from .api import APIAuthenticationError, APIConnectionDeviceNotFoundError
from .coordinator import VogelsMotionMountBleCoordinator
from .services import async_setup_services

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [
    Platform.BINARY_SENSOR,
    Platform.BUTTON,
    Platform.NUMBER,
    Platform.SELECT,
    Platform.SENSOR,
    Platform.SWITCH,
    Platform.TEXT,
]

type VogelsMotionMountBleConfigEntry = ConfigEntry[VogelsMotionMountBleCoordinator]

# for this version read_only was added to config selectors
MIN_HA_VERSION = "2025.6.0"

if version.parse(ha_version) < version.parse(MIN_HA_VERSION):
    raise RuntimeError(
        f"Vogels Motion Mount BLE requires Home Assistant {MIN_HA_VERSION}+"
    )


async def async_setup(
    hass: HomeAssistant, entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount integration services."""
    _LOGGER.debug("async_setup called with config_entry: %s", entry)
    async_setup_services(hass)
    return True


async def async_setup_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount Integration from a config entry."""
    _LOGGER.debug("async_setup_entry called with config_entry: %s", config_entry)

    # Registers update listener to update config entry when options are updated.
    unsub_update_listener = config_entry.add_update_listener(async_reload_entry)

    # Initialise the coordinator that manages data updates from your api.
    coordinator = VogelsMotionMountBleCoordinator(
        hass=hass,
        config_entry=config_entry,
        unsub_update_listener=unsub_update_listener,
    )

    try:
        await coordinator.api.refresh_data()
    except APIConnectionDeviceNotFoundError as err:
        raise ConfigEntryNotReady("error_device_not_found") from err
    except APIAuthenticationError as err:
        raise ConfigEntryAuthFailed("error_invalid_athentication") from err
    except Exception as err:
        raise ConfigEntryError(
            translation_key=f"Something went wrong {err}",
            translation_placeholders={"error": err},
        ) from err

    config_entry.runtime_data = coordinator
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
    return True


async def async_reload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> None:
    """Reload config entry."""
    _LOGGER.debug(
        "async_reload_entry async_reload with pin %s", config_entry.data["conf_pin"]
    )
    await hass.config_entries.async_reload(config_entry.entry_id)


async def async_unload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Unload a config entry."""
    _LOGGER.debug("async_unload_entry")
    if unload_ok := await hass.config_entries.async_unload_platforms(
        config_entry, PLATFORMS
    ):
        _LOGGER.debug("async_unload_entry pop")
        coordinator: VogelsMotionMountBleCoordinator = config_entry.runtime_data
        await coordinator.unload()

    return unload_ok
