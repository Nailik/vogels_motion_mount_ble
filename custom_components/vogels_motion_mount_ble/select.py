"""Select entities to define properties for Vogels Motion Mount BLE entities."""

from homeassistant.components.select import SelectEntity
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from . import VogelsMotionMountBleConfigEntry
from .api import (
    SettingsRequestType,
    VogelsMotionMountActionType,
    VogelsMotionMountAutoMoveType,
)
from .base import VogelsMotionMountBleBaseEntity
from .coordinator import VogelsMotionMountBleCoordinator


async def async_setup_entry(
    _: HomeAssistant,
    config_entry: VogelsMotionMountBleConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the Selectors for automove."""
    coordinator: VogelsMotionMountBleCoordinator = config_entry.runtime_data

    async_add_entities(
        [
            AutomoveSelect(coordinator),
            FreezePresetSelect(coordinator),
        ]
    )


class AutomoveSelect(VogelsMotionMountBleBaseEntity, SelectEntity):
    """Implementation of the Automove Selector."""

    _attr_unique_id = "auto_move"
    _attr_translation_key = _attr_unique_id
    _attr_options = ["off", "hdmi_1", "hdmi_2", "hdmi_3", "hdmi_4", "hdmi_5"]
    _attr_icon = "mdi:autorenew"

    @property
    def available(self) -> bool:  # type: ignore
        """Set availability if preset exists and user has permission."""
        return self.coordinator.api.has_permission(
            action_type=VogelsMotionMountActionType.Settings,
            settings_request_type=SettingsRequestType.change_tv_on_off_detection,
        )

    @property
    def current_option(self):  # type: ignore
        """Return the current active automove option."""
        if self.coordinator.data.automove_type is None:
            return None
        return self.coordinator.data.automove_type.value

    async def async_select_option(self, option: str) -> None:
        """Select an option."""
        await self.coordinator.api.set_automove(option)


class FreezePresetSelect(VogelsMotionMountBleBaseEntity, SelectEntity):
    """Implementation of the Freeze preset Selector."""

    _attr_unique_id = "freeze_preset"
    _attr_translation_key = _attr_unique_id
    _attr_icon = "mdi:snowflake"

    @property
    def current_option(self):  # type: ignore
        """Return the current selected freeze preset."""
        if self.coordinator.data.freeze_preset_index is None:
            return None
        return self.options[self.coordinator.data.freeze_preset_index]

    @property
    def options(self) -> list[str]:  # type: ignore
        """Return the possible options."""
        # Dynamically generated based on coordinator data
        return ["0"] + [
            str(preset.name)
            for _, preset in self.coordinator.data.presets.items()
            if preset is not None
        ]

    @property
    def available(self) -> bool:  # type: ignore
        """Set availability if automove is turned on."""
        if (
            self.coordinator.data
            and self.coordinator.data.automove_type
            is not VogelsMotionMountAutoMoveType.Off
            and self.coordinator.api.has_permission(
                action_type=VogelsMotionMountActionType.Settings,
                settings_request_type=SettingsRequestType.change_default_position,
            )
        ):
            return True
        return False

    async def async_select_option(self, option: str) -> None:
        """Select an option."""
        index = self._attr_options.index(option)
        await self.coordinator.api.set_freeze_preset(index)
