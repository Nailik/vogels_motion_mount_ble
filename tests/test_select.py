"""Tests for select entities."""

from unittest.mock import AsyncMock, patch

import pytest
from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    snapshot_platform,
)
from syrupy.assertion import SnapshotAssertion

from custom_components.vogels_motion_mount_ble.coordinator import (
    VogelsMotionMountBleCoordinator,
)
from custom_components.vogels_motion_mount_ble.data import (
    VogelsMotionMountAutoMoveType,
    VogelsMotionMountPreset,
    VogelsMotionMountPresetData,
)
from custom_components.vogels_motion_mount_ble.select import (
    AutomoveSelect,
    FreezePresetSelect,
)
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from .conftest import setup_integration  # noqa: TID251

# -------------------------------
# region Setup
# -------------------------------


async def test_all_entities(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test all entities."""
    with patch(
        "custom_components.vogels_motion_mount_ble.PLATFORMS", [Platform.SELECT]
    ):
        await setup_integration(hass, mock_config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)


# -------------------------------
# region Actions
# -------------------------------


@pytest.mark.asyncio
async def test_automove_select_option_zero(mock_coord: VogelsMotionMountBleCoordinator):
    """Test automove select option."""
    mock_coord.set_automove = AsyncMock()
    # Current automove → Hdmi_2_On (value=4)
    mock_coord.data.automove = VogelsMotionMountAutoMoveType.Hdmi_2_On
    select = AutomoveSelect(mock_coord)

    await select.async_select_option("0")

    # "0" means Off → should pick Hdmi_2_Off (value=5)
    mock_coord.set_automove.assert_awaited_once_with(
        VogelsMotionMountAutoMoveType.Hdmi_2_Off
    )


@pytest.mark.asyncio
async def test_automove_select_option_enabled(
    mock_coord: VogelsMotionMountBleCoordinator,
):
    """Test automove select option."""
    mock_coord.set_automove = AsyncMock()
    mock_coord.data.automove = VogelsMotionMountAutoMoveType.Hdmi_1_Off
    select = AutomoveSelect(mock_coord)

    await select.async_select_option("3")

    # "3" means Hdmi_3_On (value=8)
    mock_coord.set_automove.assert_awaited_once_with(
        VogelsMotionMountAutoMoveType.Hdmi_3_On
    )


@pytest.mark.asyncio
async def test_freeze_preset_select_option(mock_coord: VogelsMotionMountBleCoordinator):
    """Test freeze preset option."""
    mock_coord.set_freeze_preset = AsyncMock()
    # Build presets
    presets = [
        VogelsMotionMountPreset(
            0, VogelsMotionMountPresetData(name="one", distance=10, rotation=5)
        ),
        VogelsMotionMountPreset(
            1, VogelsMotionMountPresetData(name="two", distance=20, rotation=10)
        ),
    ]
    mock_coord.data.presets = presets
    select = FreezePresetSelect(mock_coord)

    # Options will be ["0", "one", "two"]
    await select.async_select_option("two")

    # "two" is index 2
    mock_coord.set_freeze_preset.assert_awaited_once_with(2)


# -------------------------------
# region Current option
# -------------------------------


def test_automove_select_current_option_off(
    mock_coord: VogelsMotionMountBleCoordinator,
):
    """Test automove off option."""
    # Off case: Hdmi_3_Off = 9 → odd → maps to "0"
    mock_coord.data.automove = VogelsMotionMountAutoMoveType.Hdmi_3_Off
    select = AutomoveSelect(mock_coord)
    assert select.current_option == "0"


def test_automove_select_current_option_on(mock_coord: VogelsMotionMountBleCoordinator):
    """Test automove on option."""
    # On case: Hdmi_4_On = 12 → even → maps to (12 // 4) + 1 = 4
    mock_coord.data.automove = VogelsMotionMountAutoMoveType.Hdmi_4_On
    select = AutomoveSelect(mock_coord)
    assert select.current_option == "4"


def test_freeze_preset_select_current_option_none_index(
    mock_coord: VogelsMotionMountBleCoordinator,
):
    """Test freeze preset no option."""
    mock_coord.data.freeze_preset_index = None
    mock_coord.data.presets = []
    select = FreezePresetSelect(mock_coord)
    assert select.current_option is None


def test_freeze_preset_select_current_option_invalid_index(
    mock_coord: VogelsMotionMountBleCoordinator,
):
    """Test freeze preset invalid index."""
    mock_coord.data.freeze_preset_index = 99  # out of range
    mock_coord.data.presets = []
    select = FreezePresetSelect(mock_coord)
    assert select.current_option is None


def test_freeze_preset_select_current_option_valid(
    mock_coord: VogelsMotionMountBleCoordinator,
):
    """Test freeze preset valid option."""
    presets = [
        VogelsMotionMountPreset(
            0, VogelsMotionMountPresetData(name="one", distance=10, rotation=5)
        ),
        VogelsMotionMountPreset(
            1, VogelsMotionMountPresetData(name="two", distance=20, rotation=10)
        ),
    ]
    mock_coord.data.presets = presets
    mock_coord.data.freeze_preset_index = 2  # corresponds to "two"
    select = FreezePresetSelect(mock_coord)

    # Options should be ["0", "one", "two"]
    assert select.current_option == "two"
