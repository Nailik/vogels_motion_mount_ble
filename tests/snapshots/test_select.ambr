# serializer version: 1
# name: test_all_entities[select.mount_automove_setting-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        '0',
        '1',
        '2',
        '3',
        '4',
        '5',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'select.mount_automove_setting',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:autorenew',
    'original_name': 'Automove setting',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'auto_move',
    'unique_id': 'auto_move',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[select.mount_automove_setting-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Automove setting',
      'icon': 'mdi:autorenew',
      'options': list([
        '0',
        '1',
        '2',
        '3',
        '4',
        '5',
      ]),
    }),
    'context': <ANY>,
    'entity_id': 'select.mount_automove_setting',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1',
  })
# ---
# name: test_all_entities[select.mount_freeze_preset-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'options': list([
        '0',
        'Movie Mode',
        'Movie Mode',
        'Movie Mode',
        'Gaming Mode',
        'Movie Mode',
        'Movie Mode',
        'Movie Mode',
      ]),
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'select',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'select.mount_freeze_preset',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:snowflake',
    'original_name': 'Freeze preset',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'freeze_preset',
    'unique_id': 'freeze_preset',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[select.mount_freeze_preset-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Freeze preset',
      'icon': 'mdi:snowflake',
      'options': list([
        '0',
        'Movie Mode',
        'Movie Mode',
        'Movie Mode',
        'Gaming Mode',
        'Movie Mode',
        'Movie Mode',
        'Movie Mode',
      ]),
    }),
    'context': <ANY>,
    'entity_id': 'select.mount_freeze_preset',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '0',
  })
# ---
